contracts/EscrowedNori.sol

```diff
@@ -0,0 +1,725 @@
+    private _addressToEscrowSchedules;
+
+  /**
+   * @notice The BridgedPolygonNORI contract that this contract wraps tokens for
+   */
+  BridgedPolygonNORI private _bridgedPolygonNori;
+
+  /**
+   * @notice The Removal contract that accounts for supply.
+   */
+  Removal private _removal;
+
+  /**
+   * @notice The [ERC-1820](https://eips.ethereum.org/EIPS/eip-1820) pseudo-introspection registry
+   * contract
+   *
+   * @dev Registering that EscrowedNORI implements the ERC777TokensRecipient interface with the registry is a
+   * requiremnt to be able to receive ERC-777 BridgedPolygonNORI tokens. Once registered, sending BridgedPolygonNORI
+   * tokens to this contract will trigger tokensReceived as part of the lifecycle of the BridgedPolygonNORI transaction
+   */
+  IERC1820RegistryUpgradeable private _erc1820;
+
+  /**
+   * @notice Emitted on successful creation of a new escrow schedule.
+   */
+  event EscrowScheduleCreated(
+    address indexed recipient,
+    uint256 indexed startTime
+  );
+
+  /**
+   * @notice Emitted on when unreleased tokens of an active escrow schedule are revoked.
+   */
+  event UnreleasedTokensRevoked(
+    uint256 indexed atTime,
+    uint256 indexed removalId,
+    uint256 indexed escrowScheduleId,
+    uint256 quantity
+  );
+
+  /**
+   * @notice Emitted on withdwal of fully unlocked tokens.
+   */
+  event TokensClaimed(
+    address indexed from,
+    address indexed to,
+    uint256 quantity
+  );
+
+  // todo document expected initialzation state
+  // todo I switched the visibility of this function from public to external...
+  // is that right for an initializer or was there a reason it was public?
+  function initialize(
+    BridgedPolygonNORI bridgedPolygonNoriAddress,
+    Removal removalAddress
+  ) external initializer {
+    address[] memory operators = new address[](1);
+    operators[0] = _msgSender();
+    __Context_init_unchained();
+    __ERC165_init_unchained();
+    __AccessControl_init_unchained();
+    __AccessControlEnumerable_init_unchained();
+    __Pausable_init_unchained();
+    __ERC777PresetPausablePermissioned_init_unchained();
+    __ERC777_init_unchained("Escrowed NORI", "eNORI", operators);
+    _bridgedPolygonNori = bridgedPolygonNoriAddress;
+    _removal = removalAddress;
+    _ERC1820_REGISTRY.setInterfaceImplementer(
+      address(this),
+      ERC777_TOKENS_RECIPIENT_HASH,
+      address(this)
+    );
+    _grantRole(ESCROW_CREATOR_ROLE, _msgSender());
+  }
+
+  /**
+   * @notice This function is triggered when BridgedPolygonNORI is sent to this contract
+   *
+   * @dev Sending BridgedPolygonNORI to this contract triggers the tokensReceived hook defined by the ERC-777 standard
+   * because this contract is a registered ERC777 tokens recipient.
+   *
+   * [See here for more](
+   * https://github.com/ethereum/EIPs/blob/master/EIPS/eip-777.md#erc777tokensrecipient-and-the-tokensreceived-hook)
+   */
+  function tokensReceived(
```

Comment:
:sweat_smile:  really wish we had made a shared WrappedERC777 token base contract since so much of this logic is similar/extended to LockedNORI

Reply:
mmm yeah... if we'd been able to develop this at the same time as `LockedNORI` we could have had that foresight.  thought fwiw, the body of all of these functions is actually quite different between the two contracts.  We could have defined some sort of `WrappedERC777` interface but we'd be the only ones using it...